File: algorithm-questions/dynamicProgramming/min-cost-path/MinCostPath.java
public class MinCostPath
{
	static final int R = 3,  C = 3;
	static int dp[][] = new int[R][C]; 
	static int a[][] = 
	{
			{1, 2, 3},
			{4, 8, 2},
			{1, 5, 3}
	};
	static 
	{
		for (int i = 0; i < R; i++)
		{
			for (int j = 0; j < C; j++)
			{
				dp[i][j] = -1;
			}
		}
	}
	static int solution(int a[][], int i, int j)
	{
		if (i < 0 || j < 0)
		{
			return Integer.MAX_VALUE;
		}
		if (i == 0 && j == 0)
		{
			return a[0][0];
		}
		if (dp[i][j] != -1)
		{
			return dp[i][j];
		}
		return dp[i][j] = (a[i][j] + min(solution(a, i-1, j), solution(a, i, j-1), solution(a, i-1, j-1)));
	}
	
	public static int min(int a, int b, int c)
	{
		if (a < b)
		{
			return a < c ? a : c;
		}
		else
		{
			return b < c ? b : c;
		}
	}
	
	public static void main(String[] args)
	{
		
		System.out.println(solution(a, 2, 2));
	}
}

File: algorithm-questions/dynamicProgramming/min-cost-path/MinCostPath.cpp
#include <stdio.h>
#include <iostream>
#include <bits/stdc++.h>

#define R 3
#define C 3
int dp[R][C];
int a[R][C];

void _initial()
{
	for (int x = 0; x < R;x++)
	{
		for (int y = 0; y < C; y++)
		{
			dp[x][y] = -1;
		}
	}
}

int min(int a, int b, int c)
{
	if (a < b)
	{
		return a < c ? a : c;
	}
	else
	{
		return b < c ? b : c;
	}
}
int solution(int arr[R][C], int i, int j)
{
	if (i < 0 || j < 0)
	{
		return INT_MAX;
	}
	if (i == 0 && j == 0)
	{
		return arr[0][0];
	}
	if (dp[i][j] != -1)
	{
		return dp[i][j];
	}
	return dp[i][j] = arr[i][j] + min(solution(arr, i, j-1), solution(arr, i-1, j), solution(arr, i-1, j-1));
}


int main()
{
	_initial();
	int a[R][C] = {{1, 2, 3},
			  {4, 8, 2},
			  {1, 5, 3}};
	std::cout << solution(a, 2,2);
	return 0;
}
